File: go/utility/prepare_auth.go
package utility

import (
	"encoding/base64"

	vs "github.com/voxgig-sdk/tangocard-sdk/go/utility/struct"

	"github.com/voxgig-sdk/tangocard-sdk/go/core"
)

const credName = "authorization"
const optionApikey = "apikey"
const optionSecret = "secret"
const notFound = "__NOTFOUND__"

func prepareAuthUtil(ctx *core.Context) (*core.Spec, error) {
	spec := ctx.Spec
	if spec == nil {
		return nil, ctx.MakeError("auth_no_spec",
			"Expected context spec property to be defined.")
	}

	headers := spec.Headers
	options := ctx.Client.OptionsMap()

	// Public APIs that need no auth omit the options.auth block entirely.
	if options["auth"] == nil {
		delete(headers, credName)
		return spec, nil
	}

	apikey := vs.GetProp(options, optionApikey, notFound)

	skip := false
	if apikey == nil {
		skip = true
	} else if apikeyStr, ok := apikey.(string); ok &&
		(apikeyStr == notFound || apikeyStr == "") {
		skip = true
	}

	// True HTTP Basic Auth needs TWO credentials, base64-joined - a single
	// token in the header (the branch below) can never authenticate against
	// an API that actually checks `Authorization: Basic base64(user:pass)`.
	if basicAuth, _ := vs.GetPath(options, []any{"auth", "basic"}).(bool); basicAuth {
		secret := vs.GetProp(options, optionSecret, notFound)

		noSecret := false
		if secret == nil {
			noSecret = true
		} else if secretStr, ok := secret.(string); ok &&
			(secretStr == notFound || secretStr == "") {
			noSecret = true
		}

		if skip || noSecret {
			delete(headers, credName)
		} else {
			apikeyVal, _ := apikey.(string)
			secretVal, _ := secret.(string)
			b64 := base64.StdEncoding.EncodeToString([]byte(apikeyVal + ":" + secretVal))

			basicPrefix := ""
			if ap := vs.GetPath(options, []any{"auth", "prefix"}); ap != nil {
				basicPrefix, _ = ap.(string)
			}
			// Empty prefix (raw apiKey credential) must not add a leading space.
			if basicPrefix == "" {
				headers[credName] = b64
			} else {
				headers[credName] = basicPrefix + " " + b64
			}
		}

		return spec, nil
	}

	if skip {
		delete(headers, credName)
	} else {
		authPrefix := ""
		if ap := vs.GetPath(options, []any{"auth", "prefix"}); ap != nil {
			authPrefix, _ = ap.(string)
		}
		apikeyVal := ""
		if av, ok := apikey.(string); ok {
			apikeyVal = av
		}
		// Empty prefix (raw apiKey credential) must not add a leading space.
		if authPrefix == "" {
			headers[credName] = apikeyVal
		} else {
			headers[credName] = authPrefix + " " + apikeyVal
		}
	}

	return spec, nil
}

File: php/utility/PrepareAuth.php
<?php
declare(strict_types=1);

// Tangocard SDK utility: prepare_auth

class TangocardPrepareAuth
{
    private const HEADER_AUTH = 'authorization';
    private const OPTION_APIKEY = 'apikey';
    private const OPTION_SECRET = 'secret';
    private const NOT_FOUND = '__NOTFOUND__';

    public static function call(TangocardContext $ctx): array
    {
        $spec = $ctx->spec;
        if (!$spec) {
            return [null, $ctx->make_error('auth_no_spec', 'Expected context spec property to be defined.')];
        }

        $headers = &$spec->headers;
        $options = $ctx->client->options_map();

        // Public APIs that need no auth omit the options.auth block entirely.
        if (!isset($options['auth']) || $options['auth'] === null) {
            unset($headers[self::HEADER_AUTH]);
            return [$spec, null];
        }

        $apikey = \Voxgig\Struct\Struct::getprop($options, self::OPTION_APIKEY, self::NOT_FOUND);

        // True HTTP Basic Auth needs TWO credentials, base64-joined - a
        // single token in the header (the branch below) can never
        // authenticate against an API that actually checks
        // `Authorization: Basic base64(user:pass)`.
        if (true === (\Voxgig\Struct\Struct::getpath($options, 'auth.basic') ?? false)) {
            $secret = \Voxgig\Struct\Struct::getprop($options, self::OPTION_SECRET, self::NOT_FOUND);
            $apikey_val = is_string($apikey) && $apikey !== self::NOT_FOUND ? $apikey : '';
            $secret_val = is_string($secret) && $secret !== self::NOT_FOUND ? $secret : '';

            if ($apikey_val === '' || $secret_val === '') {
                unset($headers[self::HEADER_AUTH]);
            } else {
                $auth_prefix = \Voxgig\Struct\Struct::getpath($options, 'auth.prefix') ?? '';
                $b64 = base64_encode("{$apikey_val}:{$secret_val}");
                $headers[self::HEADER_AUTH] = $auth_prefix === ''
                    ? $b64 : "{$auth_prefix} {$b64}";
            }

            return [$spec, null];
        }

        if (
            (is_string($apikey) && ($apikey === self::NOT_FOUND || $apikey === ''))
            || $apikey === null
        ) {
            unset($headers[self::HEADER_AUTH]);
        } else {
            $auth_prefix = \Voxgig\Struct\Struct::getpath($options, 'auth.prefix') ?? '';
            $apikey_val = is_string($apikey) ? $apikey : '';
            // Empty prefix (raw apiKey credential) must not add a leading space.
            $headers[self::HEADER_AUTH] = $auth_prefix === ''
                ? $apikey_val : "{$auth_prefix} {$apikey_val}";
        }

        return [$spec, null];
    }
}
